File: leetcode/1001-1999/1400-construct-k-palindrome-strings/ConstructKPalindromeStrings.java
public class ConstructKPalindromeStrings {
  public static void main(String[] args) {
    String s = "annabelle";
    int k = 2;

    System.out.println(new Solution().canConstruct(s, k));
  }
}

class Solution {
  public boolean canConstruct(String s, int k) {
    if (s.length() < k) {
      return false;
    }

    int totalOddCharacters = 0;

    int[] characters = new int[26];

    for (int i = 0; i < s.length(); i++) {
      characters[s.charAt(i) - 'a']++;
    }

    for (int i = 0; i < characters.length; i++) {
      totalOddCharacters += characters[i] % 2;
    }

    return totalOddCharacters <= k;
  }
}

File: leetcode/1001-2000/1401-1500/1401-1410/1400-construct-k-palindrome-strings/ConstructKPalindromeStrings.ts
function canConstruct(s: string, k: number): boolean {
  if (s.length < k) {
    return false;
  }

  let totalOddCharacters = 0;

  const characters: number[] = new Array(26).fill(0);

  const a = "a".codePointAt(0)!;

  for (let i = 0; i < s.length; i++) {
    characters[s.codePointAt(i)! - a]++;
  }

  for (let i = 0; i < characters.length; i++) {
    totalOddCharacters += characters[i] % 2;
  }

  return totalOddCharacters <= k;
};

console.log(canConstruct("annabelle", 2));

